File: RudderSDKCore/RudderSDKCore/RudderClient.m
//
//  RudderClient.m
//  RudderSDKCore
//
//  Created by Arnab Pal on 17/10/19.
//  Copyright © 2019 Rudderlabs. All rights reserved.
//

#import "RudderClient.h"
#import "EventRepository.h"
#import "RudderMessageBuilder.h"
#import "RudderElementCache.h"

static RudderClient *_instance = nil;
static EventRepository *_repository = nil;

@implementation RudderClient

+ (instancetype) getInstance {
    return _instance;
}

+ (instancetype) getInstance:(NSString *)writeKey {
    return [RudderClient getInstance:writeKey config:[[RudderConfig alloc] init]];
}

+ (instancetype) getInstance: (NSString *) writeKey config: (RudderConfig*) config {
    if (_instance == nil) {
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            _instance = [[self alloc] init];
            _repository = [EventRepository initiate:writeKey config:config];
        });
    }
    return _instance;
}

- (void)trackMessage:(RudderMessage *)message {
    if (_repository != nil && message != nil) {
        message.type = @"track";
        [_repository dump:message];
    }
}

- (void)trackWithBuilder:(RudderMessageBuilder *)builder{
    [self trackMessage:[builder build]];
}

- (void)track:(NSString *)eventName {
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    [builder setEventName:eventName];
    [self trackMessage:[builder build]];
}

- (void)track:(NSString *)eventName properties:(NSDictionary<NSString *,NSObject *> *)properties {
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    [builder setEventName:eventName];
    [builder setPropertyDict:properties];
    [self trackMessage:[builder build]];
}

- (void)track:(NSString *)eventName properties:(NSDictionary<NSString *,NSObject *> *)properties options:(RudderOption *)options {
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    [builder setEventName:eventName];
    [builder setPropertyDict:properties];
    [builder setRudderOption:options];
    [self trackMessage:[builder build]];
}

- (void) screenWithMessage:(RudderMessage *)message {
    if (_repository != nil && message != nil) {
        message.type = @"screen";
        [_repository dump:message];
    }
}

- (void)screenWithBuilder:(RudderMessageBuilder *)builder {
    [self screenWithMessage:[builder build]];
}

- (void)screen:(NSString *)screenName {
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    NSMutableDictionary *property = [[NSMutableDictionary alloc] init];
    [property setValue:screenName forKey:@"name"];
    [builder setEventName:screenName];
    [builder setPropertyDict:property];
    [self screenWithMessage:[builder build]];
}

- (void)screen:(NSString *)screenName properties:(NSDictionary<NSString *,NSObject *> *)properties {
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    [builder setEventName:screenName];
    [builder setPropertyDict:properties];
    [self screenWithBuilder:builder];
}

- (void)screen:(NSString *)screenName properties:(NSDictionary<NSString *,NSObject *> *)properties options:(RudderOption *)options {
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    [builder setEventName:screenName];
    [builder setPropertyDict:properties];
    [builder setRudderOption:options];
    [self screenWithBuilder:builder];
}

- (void)group:(NSString *)groupId{
    
}

- (void)group:(NSString *)groupId traits:(NSDictionary<NSString *,NSObject *> *)traits {
    
}

- (void)group:(NSString *)groupId traits:(NSDictionary<NSString *,NSObject *> *)traits options:(NSDictionary<NSString *,NSObject *> *)options {
    
}

- (void)alias:(NSString *)newId {
    
}

- (void)alias:(NSString *)newId options:(NSDictionary<NSString *,NSObject *> *)options {
    
}

- (void) pageWithMessage:(RudderMessage *)message {
    if (_repository != nil && message != nil) {
        message.type = @"page";
        [_repository dump:message];
    }
}

- (void) identifyWithMessage:(RudderMessage *)message {
    if (_repository != nil && message != nil) {
        message.type = @"identify";
        NSString *userId = message.context.traits.userId;
        if (userId != nil) {
            message.userId = userId;
        }
        [_repository dump:message];
    }
}

- (void)identifyWithBuilder:(RudderMessageBuilder *)builder {
    [self identifyWithMessage:[builder build]];
}

- (void)identify:(NSString*)userId {
    RudderTraits* traitsCopy = [[RudderTraits alloc] init];
    [traitsCopy setUserId:userId];
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    [builder setEventName:@"identify"];
    [builder setUserId:userId];
    [builder setTraits:traitsCopy];
    [self identifyWithMessage:[builder build]];
}

- (void)identify:(NSString *)userId traits:(NSDictionary *)traits {
    RudderTraits* traitsObj = [[RudderTraits alloc] initWithDict: traits];
    [traitsObj setUserId:userId];
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    [builder setEventName:@"identify"];
    [builder setUserId:userId];
    [builder setTraits:traitsObj];
    [self identifyWithMessage:[builder build]];
}

- (void)identify:(NSString *)userId traits:(NSDictionary *)traits options:(NSDictionary *)options {
    RudderTraits *traitsObj = [[RudderTraits alloc] initWithDict:traits];
    [traitsObj setUserId:userId];
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    [builder setEventName:@"identify"];
    [builder setUserId:userId];
    [builder setTraits:traitsObj];
    RudderOption *optionsObj = [[RudderOption alloc] initWithDict:options];
    [builder setRudderOption:optionsObj];
    [self identifyWithMessage:[builder build]];
}

- (void)reset {
    
}

- (NSString*)getAnonymousId {
    // returns anonymousId
    return [RudderElementCache getContext].device.identifier;
}

- (RudderConfig*)configuration {
    if (_repository == nil) {
        return nil;
    }
    return [_repository getConfig];
}

+ (instancetype)sharedInstance {
    return _instance;
}

@end

File: RudderSDKCore/Classes/RudderClient.swift
//
//  RudderClient.swift
//  RudderSdkCore
//
//  Created by Arnab Pal on 09/10/19.
//  Copyright © 2019 Arnab Pal. All rights reserved.
//

import UIKit

@objcMembers public class RudderClient: NSObject {
    
    private static var instance: RudderClient? = nil;
    private var repository: EventRepository? = nil
    
    private override init() {
        // do nothing
    }
    
    private init(writeKey: String, config: RudderConfig) {
        // initiate things
        self.repository = EventRepository(_writeKey: writeKey, _config: config)
    }
    
    /*
     * API for getting instance of RudderClient with writeKey (bare minimum)
     * */
    @objc public static func getInstance(writeKey: String) -> RudderClient {
        return getInstance(writeKey: writeKey, config: RudderConfig())
    }
    
    /*
     * API for getting instance of RudderClient with config
     * */
    public static func getInstance(writeKey: String, builder: RudderConfig.Builder) -> RudderClient {
        return getInstance(writeKey: writeKey, config: builder.build())
    }
    
    /*
     * API for getting instance of RudderClient with config
     * */
    public static func getInstance(writeKey: String, config: RudderConfig) -> RudderClient {
        if (instance == nil) {
            instance = RudderClient(writeKey: writeKey, config: config)
        }
        
        return instance!
    }
    
    /*
     * package private api to be used in EventRepository
     * */
    static func getInstance() -> RudderClient? {
        return instance
    }
    
    
    /*
     * method for `track` messages
     * */
    public func track(message: RudderMessage) {
        message.setEventType(type: MessageType.TRACK)
        do {
            try self.repository?.dump(message: message)
        } catch {
            RudderLogger.logError(error: error)
        }
    }
    
    public func track(builder: RudderMessageBuilder) {
        self.track(message: builder.build())
    }
    
    /*
     * segment equivalent API
     * */
    public func track(event: String) {
        let message: RudderMessage = RudderMessageBuilder()
        .withEventName(eventName: event)
        .build()
        self.track(message: message)
    }
    
    public func track(event: String, property: Dictionary<String, NSObject>?) {
        let message: RudderMessage = RudderMessageBuilder()
        .withEventName(eventName: event)
        .withEventProperties(properties: property)
        .build()
        self.track(message: message)
    }
    
    public func track(event: String, property: Dictionary<String, NSObject>?, options: Dictionary<String, NSObject>?) {
        let message: RudderMessage = RudderMessageBuilder()
            .withEventName(eventName: event)
            .withEventProperties(properties: property)
            .withOptions(options: options)
            .build()
        self.track(message: message)
    }
    
    
    /*
     * method for `screen` messages
     * */
    public func screen(message: RudderMessage) {
        message.setEventType(type: MessageType.SCREEN)
        do {
            try self.repository?.dump(message: message)
        } catch {
            RudderLogger.logError(error: error)
        }
    }
    
    public func screen(builder: RudderMessageBuilder) {
        self.screen(message: builder.build())
    }
    
    /*
     * segment equivalent API
     * */
    public func screen(event: String) {
        let message: RudderMessage = RudderMessageBuilder()
            .withEventName(eventName: event)
            .build()
        self.screen(message: message)
    }
    
    public func screen(event: String, property: Dictionary<String, NSObject>?) {
        let message: RudderMessage = RudderMessageBuilder()
            .withEventName(eventName: event)
            .withEventProperties(properties: property)
            .build()
        self.screen(message: message)
    }
    
    public func screen(event: String, category: String, property: Dictionary<String, NSObject>?, options: Dictionary<String, NSObject>?) {
        let message: RudderMessage = RudderMessageBuilder()
            .withEventName(eventName: event)
            .withEventProperties(properties: property)
            .withEventProperty(key: "category", value: category as NSObject)
            .withOptions(options: options)
            .build()
        self.screen(message: message)
    }
    
    public func screen(event: String, property: Dictionary<String, NSObject>?, options: Dictionary<String, NSObject>?) {
        let message: RudderMessage = RudderMessageBuilder()
            .withEventName(eventName: event)
            .withEventProperties(properties: property)
            .withOptions(options: options)
            .build()
        self.screen(message: message)
    }
    
    /*
     * method for `page` messages
     * */
    public func page(message: RudderMessage) {
        message.setEventType(type: MessageType.PAGE)
        do {
            try self.repository?.dump(message: message)
        } catch {
            RudderLogger.logError(error: error)
        }
    }
    
    public func page(builder: RudderMessageBuilder) {
        self.page(message: builder.build())
    }
    
    /*
     * method for `identify` messages
     * */
    @objc public func identify(message: RudderMessage) {
        
    }
    
    @objc public func identify(builder: RudderMessageBuilder) {
        
    }
    
    @objc public func identify(traits: RudderTraits, options: AnyCodable) {
        
    }
    
    //public func identify(builder: RudderTraitsBuilder) {
        
    //}
    
    public func identify(userId: String) {
        
    }
    
    /*
     * segment equivalent API
     * */
    public func alias(event: String) {
        // to be decided
    }
    
    public func alias(event: String, options: AnyCodable) {
        // to be decided
    }
    
    public func group(groupId: String) {
        // to be decided
    }
    
    public func group(groupId: String, traits: RudderTraits) {
        // to be decided
    }
    
    public func group(groupId: String, traits: RudderTraits, options: AnyCodable?) {
        // to be decided
    }
    
    public static func setSingletonInstance(_instance: RudderClient?) {
        if (_instance != nil) {
            instance = _instance
        }
    }
    
    public func getRudderContext() -> RudderContext {
        return RudderElementCache.getCachedContext()
    }
    
    public func optOut() {
        // to be decided
    }
    
    public func reset() {

    }
    
    public func shutdown() {
        
    }
    
    public class Builder {
        private var writeKey: String
        
        public init(writeKey: String) {
            self.writeKey = writeKey
        }
        
        private var trackLifecycleEvents: Bool = false
        
        public func trackApplicationLifecycleEvents() -> Builder {
            self.trackLifecycleEvents = true
            return self
        }
        
        private var recordScreenView: Bool = false
        
        public func recordScreenViews() -> Builder {
            self.recordScreenView = true
            return self
        }
        
        private var config: RudderConfig? = nil
        
        public func withRudderConfig(config: RudderConfig) -> Builder {
            self.config = config
            return self
        }
        
        public func withRudderConfigBuilder(builder: RudderConfig.Builder) -> Builder {
            self.config = builder.build()
            return self
        }
        
        public func build() -> RudderClient {
            if (self.config == nil) {
                self.config = RudderConfig()
            }
            
            return RudderClient.getInstance(writeKey: self.writeKey, config: self.config!)
        }
    }
}
